Support Table and PartialObjectMetadata on watch

smarterclayton opened the pull request:
Clean up the code paths that lead to objects being transformed and output
with negotiation. Remove some duplicate code that was not consistent. Now,
watch will respond correctly to Table and PartialObjectMetadata requests.

/kind bug

TODO

* [x] Add unit tests
* [x] Add e2e test
* [x] Only return Table headers on the first object
* [x] During early output negotiation, filter unsupported encodings

```release-note
Watch will now support converting response objects into Table or PartialObjectMetadata forms.
```

smarterclayton committed:
Support Table and PartialObjectMetadata on watch

Clean up the code paths that lead to objects being transformed and output with negotiation.
Remove some duplicate code that was not consistent. Now, watch will respond correctly to
Table and PartialObjectMetadata requests. Add unit and integration tests.

When transforming responses to Tables, only the first watch event for a given type will
include the columns. Columns will not change unless the watch is restarted.

Add a volume attachment printer and tighten up table validation error cases.

Disable protobuf from table conversion because Tables don't have protobuf because they
use `interface{}`

smarterclayton commented:
/assign @liggitt 

First two commits are from #71542

fedebongio commented:
/cc @mbohlool

fedebongio commented:
/cc @wenjiaswe

caesarxuchao commented:
/assign

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:
```diff
@@ -27,85 +27,67 @@ import (
-	if _, ok := result.(*metav1.Status); ok {
-		responsewriters.WriteObject(statusCode, scope.Kind.GroupVersion(), scope.Serializer, result, w, req)
-		return
+// transformObject takes the object as returned by storage and
```
Broken comment.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:
```diff
@@ -27,85 +27,67 @@ import (
+	}
+}
+
+// serializerForTransform returns the appropriate serializer for the input media type
```
mismatching comment

smarterclayton commented:
Added integration and unit tests that should cover the primary paths through watch.  Removing `WIP`, I think this is mostly complete.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:
```diff
@@ -201,3 +129,91 @@ func (e errNotAcceptable) Status() metav1.Status {
+
+func asV1Beta1PartialObjectMetadata(result runtime.Object) (runtime.Object, error) {
+	if meta.IsListType(result) {
+		// TODO: this should be calculated earlier
```
no, removed

smarterclayton commented:
Should be ready for review

liggitt commented on pkg/printers/humanreadable.go:
```diff
@@ -509,14 +509,17 @@ func (h *HumanReadablePrinter) PrintTable(obj runtime.Object, options PrintOptio
-				continue
+	var columns []metav1beta1.TableColumnDefinition
+	if !options.NoHeaders {
+		columns = handler.columnDefinitions
```
we don't need the column definitions for anything non-display-related?

smarterclayton replied:
In the code path that results in a table being created, no.  If you somehow always wired this to true and sent it to the server, `kubectl` would not correctly handle priority.

However, for all non-watch cases getting no headers is assumed you know what you are asking for (tabular data without headers), while in watch cases we define the behavior explicitly.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1beta1/types.go:
```diff
@@ -133,6 +133,10 @@ const (
 	v1.TypeMeta `json:",inline"`
+
+	// NoHeaders is only exposed for internal callers.
+	NoHeaders bool `json:"-"`
```
o_O

smarterclayton replied:
Other option includes passing internal around everywhere but we can't really do that on some of the calls easily.  I couldn't think of anything better...

smarterclayton commented on pkg/registry/storage/volumeattachment/storage/storage.go:
```diff
@@ -50,6 +53,8 @@ func NewStorage(optsGetter generic.RESTOptionsGetter) *VolumeAttachmentStorage {
 		DeleteStrategy:      volumeattachment.Strategy,
 		ReturnDeletedObject: true,
+
+		TableConvertor: printerstorage.TableConvertor{TablePrinter: printers.NewTablePrinter().With(printersinternal.AddHandlers)},
```
This package wasn't loaded when the reflexive test ran.

smarterclayton commented on pkg/printers/storage/storage.go:
```diff
@@ -29,5 +29,9 @@ type TableConvertor struct {
 func (c TableConvertor) ConvertToTable(ctx context.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1beta1.Table, error) {
-	return c.TablePrinter.PrintTable(obj, printers.PrintOptions{Wide: true})
+	noHeaders := false
+	if opts, ok := tableOptions.(*metav1beta1.TableOptions); ok && opts.NoHeaders {
```
Switch with an error case?

smarterclayton commented:
Updated with two of the comments - I'm open to options on what to do with non-internal Options.

smarterclayton commented:
pc load letter

/retest

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -106,10 +120,17 @@ func serveWatch(watcher watch.Interface, scope RequestScope, req *http.Request,
+				return obj
+			}
+			if tableOptions, ok := options.(*metav1beta1.TableOptions); ok {
+				tableOptions.NoHeaders = true
```
this sets NoHeaders=true after rendering the first object, so headers on subsequent objects are not sent? bears a comment

smarterclayton replied:
correct, will comment

liggitt replied:
is mutating `options` safe? do we need to check if `tableOptions` is non-nil?

liggitt commented:
ran into issues trying to use this with streaming proto tables:

```
$ curl 'http://localhost:8080/api/v1/namespaces?watch=true&includeObject=Metadata' -H "Accept: application/vnd.kubernetes.protobuf;as=Table;v=v1beta1;g=meta.k8s.io" -v
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 8080 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /api/v1/namespaces?watch=true&includeObject=Metadata HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: application/vnd.kubernetes.protobuf;as=Table;v=v1beta1;g=meta.k8s.io
> 
< HTTP/1.1 200 OK
< Content-Type: application/vnd.kubernetes.protobuf;stream=watch
< Date: Wed, 06 Mar 2019 22:50:13 GMT
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
```

no output

server log:
```
E0306 17:51:37.487934   44702 watch.go:233] unable to encode watch object *v1beta1.Table: object *v1beta1.Table does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message
```

needs integration tests to exercise Table and PartialObjectMetadata with protobug streaming, and also to ensure we write an error to the output stream if an encoding error is encountered

smarterclayton commented:
Tables don't have proto because of table cells - we never agreed on a serialization and both deepcopy-gen and openapi-gen have bugs that prevent the naive fix (serialize JSON for each cell and store in proto as a string) from working.  Given timeline, this should block GA but doesn't have to block fixing this, added an error condition to disable.

I added test cases for PartialObjectMetadata which has no such problems,  but that reminded me that to test watch deserialization we need #73937

smarterclayton commented:
/retest

liggitt commented:
convertor#ConvertToTable also sets ColumnDefinitions and should honor tableOptions.NoHeaders:

```
func (c *convertor) ConvertToTable(ctx context.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1beta1.Table, error) {
	table := &metav1beta1.Table{
		ColumnDefinitions: c.headers,
	}
```

spiffxp commented:
/hold
v1.14 release lead here, this is a size/XXL PR that is attempting to land right before Code Freeze... are you _really sure_ this is a good idea?

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/apiserver_test.go:
```diff
@@ -1918,6 +1920,152 @@ func TestGetTable(t *testing.T) {
+		{
+			item:   true,
+			accept: runtime.ContentTypeJSON + ";as=Table;v=v1beta1;g=meta.k8s.io",
+			expected: []*metav1.WatchEvent{
```
the test loop never looks at this field...

smarterclayton replied:
fixed, as well as the other conversion.  Tests roughly the same thing integration already did, at a lower level.

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -61,42 +61,56 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
-func serveWatch(watcher watch.Interface, scope RequestScope, req *http.Request, w http.ResponseWriter, timeout time.Duration) {
-	// negotiate for the stream serializer
-	serializer, err := negotiation.NegotiateOutputStreamSerializer(req, scope.Serializer)
+func serveWatch(watcher watch.Interface, scope RequestScope, mediaTypeOptions negotiation.MediaTypeOptions, req *http.Request, w http.ResponseWriter, timeout time.Duration) {
```
there's a lot of stuff in mediaTypeOptions, and we're ignoring almost all of it. we pull media type from serializer like this:
```
// TODO: next step, get back mediaTypeOptions from negotiate and return the exact value here
mediaType := serializer.MediaType
if mediaType != runtime.ContentTypeJSON {
	mediaType += ";stream=watch"
}
```
instead of using `mediaTypeOptions.Accepted`

we also don't check if `mediaTypeOptions.Accepted.Serializer` is set

is there a way to get the information we need more crisply?

smarterclayton replied:
Serializer is always set - it's not allowed to be. nil.

smarterclayton replied:
And this doesn't change the negotiation that we had before (in terms of behavior relative to 1.13).

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -61,42 +61,56 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+		var embedded runtime.Serializer
+		for _, supported := range contentSerializer.SupportedMediaTypes() {
+			if supported.MediaType == serializer.MediaType {
+				embedded = supported.Serializer
```
break?

smarterclayton commented:
Added a CRD table test and rebased, I can take your patch for partial negotiation.

liggitt commented:
on master, this works:

```sh
curl 'http://localhost:8080/api/v1/namespaces' -H "Accept: application/yaml;as=Table;v=v1beta1;g=meta.k8s.io"
```

```yaml
apiVersion: meta.k8s.io/v1beta1
kind: Table
metadata:
  resourceVersion: "327"
  selfLink: /api/v1/namespaces
columnDefinitions:
- description: 'Name must be unique within a namespace. Is required when creating
    resources, although some resources may allow a client to request the generation
    of an appropriate name automatically. Name is primarily intended for creation
    idempotence and configuration definition. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/identifiers#names'
  format: name
  name: Name
  priority: 0
  type: string
- description: The status of the namespace
  format: ""
  name: Status
  priority: 0
  type: string
- description: |-
    CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC.

    Populated by the system. Read-only. Null for lists. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
  format: ""
  name: Age
  priority: 0
  type: string
rows:
...
```

on this branch:
```
curl 'http://localhost:8080/api/v1/namespaces' -H "Accept: application/yaml;as=Table;v=v1beta1;g=meta.k8s.io"

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "only the following media types are accepted when converting to meta.k8s.io/v1beta1, Kind=Table: application/json",
  "reason": "NotAcceptable",
  "code": 406
}
```

for non-watch, we need coverage of the content types we currently support

liggitt commented:
> I can take your patch for partial negotiation.

yeah, I updated it at https://github.com/liggitt/kubernetes/commits/not-acceptable

smarterclayton commented:
/retest

smarterclayton commented:
/retest

liggitt commented:
This lgtm. I'm inclined to hold it for 1.15 at this point.

liggitt commented:
/lgtm
/approve
/milestone v1.15

liggitt commented:
/remove-label api-review

liggitt commented:
/hold cancel

liggitt commented:
does this PR also fix the issue https://github.com/kubernetes/kubernetes/pull/62175 was trying to address?

smarterclayton commented:
As commented on slack, there is a separate PR that resolves it that
simplifies the client code - Fix watch negotiation when using a non-default
mime type in the client
<https://github.com/kubernetes/kubernetes/pull/73938>

I want to look as well at fixing the CRD clients to bypass protobuf if the
provided object can't be encoded into protobuf on update operations, which
would allow us to default to protobuf clients.

On Tue, Mar 19, 2019 at 11:01 PM fejta-bot <notifications@github.com> wrote:

> /retest
> This bot automatically retries jobs that failed/flaked on approved PRs
> (send feedback to fejta <https://github.com/fejta>).
>
> Review the full test history <https://gubernator.k8s.io/pr/71548> for
> this PR.
>
> Silence the bot with an /lgtm cancel or /hold comment for consistent
> failures.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/71548#issuecomment-474665388>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p2XrIDn-Ikg7hLXzkw0oigKLgb3iks5vYaR2gaJpZM4Y48S1>
> .
>
